File: problems/173/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> st;
    TreeNode *root = nullptr;
public:
    BSTIterator(TreeNode* _root): root{_root} {}
    
    int next() {
        int ans = -1;
        
        if (not empty(st) or root) {
            while (root) {
                st.push(root);
                root = root->left;
            }
            
            root = st.top();
            st.pop();
            
            ans = root->val;
            root = root->right;
        }
        
        return ans;
    }
    
    bool hasNext() {
        return not empty(st) or root;
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: problems/173/solution.go
package main
import lls "github.com/emirpasic/gods/stacks/linkedliststack"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type BSTIterator struct {
    st *lls.Stack
    root *TreeNode
}


func Constructor(root *TreeNode) BSTIterator {
    return BSTIterator{
        lls.New(),
        root,
    }
}


func (this *BSTIterator) Next() int {
    ans := 0
    
    for !this.st.Empty() || this.root != nil {
        if this.root != nil {
            this.st.Push(this.root)
            this.root = this.root.Left
        } else {
            res, _ := this.st.Pop()
            node := res.(*TreeNode)
            ans,this.root = node.Val, node.Right
            break
        }
    }
    
    return ans
}


func (this *BSTIterator) HasNext() bool {
    return !this.st.Empty() || this.root != nil
}


/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */